File: app/src/main/java/com/benny/openlauncher/activity/MinibarEditActivity.java
package com.benny.openlauncher.activity;

import android.os.Bundle;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SwitchCompat;
import android.support.v7.widget.Toolbar;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.benny.openlauncher.R;
import com.benny.openlauncher.util.AppSettings;
import com.benny.openlauncher.util.LauncherAction;
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter;
import com.mikepenz.fastadapter.items.AbstractItem;
import com.mikepenz.fastadapter_extensions.drag.ItemTouchCallback;
import com.mikepenz.fastadapter_extensions.drag.SimpleDragCallback;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MinibarEditActivity extends ColorActivity implements ItemTouchCallback {
    @BindView(R.id.toolbar)
    Toolbar _toolbar;
    @BindView(R.id.enableSwitch)
    SwitchCompat _enableSwitch;
    @BindView(R.id.recyclerView)
    RecyclerView _recyclerView;
    private FastItemAdapter<Item> _adapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_minibar_edit);
        ButterKnife.bind(this);
        setSupportActionBar(_toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        setTitle(R.string.minibar);

        _adapter = new FastItemAdapter<>();

        SimpleDragCallback touchCallback = new SimpleDragCallback(this);
        ItemTouchHelper touchHelper = new ItemTouchHelper(touchCallback);
        touchHelper.attachToRecyclerView(_recyclerView);

        _recyclerView.setLayoutManager(new LinearLayoutManager(this));
        _recyclerView.setAdapter(_adapter);

        final ArrayList<LauncherAction.ActionDisplayItem> minibarArrangement = AppSettings.get().getMinibarArrangement();
        for (LauncherAction.ActionDisplayItem item : LauncherAction.actionDisplayItems) {
            _adapter.add(new Item(item, minibarArrangement.contains(item)));
        }

        boolean minibarEnable = AppSettings.get().getMinibarEnable();
        _enableSwitch.setChecked(minibarEnable);
        _enableSwitch.setText(minibarEnable ? R.string.on : R.string.off);
        _enableSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                buttonView.setText(isChecked ? R.string.on : R.string.off);
                AppSettings.get().setMinibarEnable(isChecked);
                if (HomeActivity.Companion.getLauncher() != null) {
                    HomeActivity.Companion.getLauncher().closeAppDrawer();
                    HomeActivity.Companion.getLauncher().getDrawerLayout().setDrawerLockMode(isChecked ? DrawerLayout.LOCK_MODE_UNLOCKED : DrawerLayout.LOCK_MODE_LOCKED_CLOSED);
                }
            }
        });

        setResult(RESULT_OK);
    }

    @Override
    protected void onPause() {
        ArrayList<String> minibarArrangement = new ArrayList<>();
        for (Item item : _adapter.getAdapterItems()) {
            if (item.enable) minibarArrangement.add(item.item._action.toString());
        }
        AppSettings.get().setMinibarArrangement(minibarArrangement);
        super.onPause();
    }

    @Override
    protected void onStop() {
        if (HomeActivity.Companion.getLauncher() != null) {
            HomeActivity.Companion.getLauncher().initMinibar();
        }
        super.onStop();
    }

    @Override
    public boolean itemTouchOnMove(int oldPosition, int newPosition) {
        Collections.swap(_adapter.getAdapterItems(), oldPosition, newPosition);
        _adapter.notifyAdapterDataSetChanged();
        return false;
    }

    @Override
    public void itemTouchDropped(int i, int i1) {
    }

    public static class Item extends AbstractItem<Item, Item.ViewHolder> {
        public final LauncherAction.ActionDisplayItem item;
        public boolean enable;

        public Item(LauncherAction.ActionDisplayItem item, boolean enable) {
            this.item = item;
            this.enable = enable;
        }

        @Override
        public int getType() {
            return 0;
        }

        @Override
        public int getLayoutRes() {
            return R.layout.item_edit_minibar;
        }

        @Override
        public ViewHolder getViewHolder(View v) {
            return new ViewHolder(v);
        }

        @Override
        public void bindView(ViewHolder holder, List payloads) {
            holder._label.setText(item._label);
            holder._description.setText(item._description);
            holder._icon.setImageResource(item._icon);
            holder._cb.setChecked(enable);
            holder._cb.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                    enable = b;
                }
            });
            super.bindView(holder, payloads);
        }

        public static class ViewHolder extends RecyclerView.ViewHolder {
            TextView _label;
            TextView _description;
            ImageView _icon;
            CheckBox _cb;

            public ViewHolder(View itemView) {
                super(itemView);
                _label = itemView.findViewById(R.id.tv);
                _description = itemView.findViewById(R.id.tv2);
                _icon = itemView.findViewById(R.id.iv);
                _cb = itemView.findViewById(R.id.cb);
            }
        }
    }
}

File: app/src/main/java/com/benny/openlauncher/activity/MinibarEditActivity.kt
package com.benny.openlauncher.activity

import android.app.Activity
import android.os.Bundle
import android.support.v4.widget.DrawerLayout
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import android.view.View
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView
import com.benny.openlauncher.R
import com.benny.openlauncher.util.AppSettings
import com.benny.openlauncher.util.LauncherAction
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter
import com.mikepenz.fastadapter.items.AbstractItem
import com.mikepenz.fastadapter_extensions.drag.ItemTouchCallback
import com.mikepenz.fastadapter_extensions.drag.SimpleDragCallback
import kotlinx.android.synthetic.main.activity_home.*
import kotlinx.android.synthetic.main.activity_minibar_edit.*
import kotlinx.android.synthetic.main.toolbar.*
import java.util.*

class MinibarEditActivity : ThemeActivity(), ItemTouchCallback {

    private var adapter: FastItemAdapter<Item>? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_minibar_edit)
        setSupportActionBar(toolbar)
        toolbar.setBackgroundColor(AppSettings.get().primaryColor)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        setTitle(R.string.minibar)

        adapter = FastItemAdapter()

        val touchCallback = SimpleDragCallback(this)
        val touchHelper = ItemTouchHelper(touchCallback)
        touchHelper.attachToRecyclerView(recyclerView)

        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = adapter

        var i = 0
        val minibarArrangement = AppSettings.get().minibarArrangement
        for (act in minibarArrangement) {
            val item = LauncherAction.getActionItemFromString(act.substring(1))
            adapter!!.add(Item(i.toLong(), item, act[0] == '0'))
            i++
        }

        val minBarEnable = AppSettings.get().minibarEnable
        enableSwitch!!.isChecked = minBarEnable
        enableSwitch!!.setText(if (minBarEnable) R.string.on else R.string.off)
        enableSwitch!!.setOnCheckedChangeListener { buttonView, isChecked ->
            buttonView.setText(if (isChecked) R.string.on else R.string.off)
            AppSettings.get().minibarEnable = isChecked
            if (CoreHome.launcher != null) {
                CoreHome.launcher?.drawer_layout?.closeDrawers()
                CoreHome.launcher?.drawer_layout?.setDrawerLockMode(if (isChecked) DrawerLayout.LOCK_MODE_UNLOCKED else DrawerLayout.LOCK_MODE_LOCKED_CLOSED)
            }
        }

        setResult(Activity.RESULT_OK)
    }

    override fun onPause() {
        val minibarArrangement = ArrayList<String>()
        for (item in adapter!!.adapterItems) {
            if (item.enable) {
                minibarArrangement.add("0" + item.item.label.toString())
            } else
                minibarArrangement.add("1" + item.item.label.toString())
        }
        AppSettings.get().minibarArrangement = minibarArrangement
        super.onPause()
    }

    override fun onStop() {
        Home.launcher?.initMinibar()
        super.onStop()
    }

    override fun itemTouchOnMove(oldPosition: Int, newPosition: Int): Boolean {
        Collections.swap(adapter!!.adapterItems, oldPosition, newPosition)
        adapter!!.notifyAdapterDataSetChanged()
        return false
    }

    override fun itemTouchDropped(i: Int, i1: Int) {}

    class Item(val id: Long, val item: LauncherAction.ActionDisplayItem, var enable: Boolean) : AbstractItem<Item, Item.ViewHolder>() {
        private var edited: Boolean = false

        override fun getType(): Int = 0

        override fun getLayoutRes(): Int = R.layout.item_minibar_edit

        override fun getViewHolder(v: View): ViewHolder = ViewHolder(v)

        override fun bindView(holder: ViewHolder, payloads: List<*>?) {
            holder.tv.text = item.label.toString()
            holder.tv2.text = item.description
            holder.iv.setImageResource(item.icon)
            holder.cb.isChecked = enable
            holder.cb.setOnCheckedChangeListener { _, b ->
                edited = true
                enable = b
            }
            super.bindView(holder, payloads)
        }

        class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            var tv: TextView = itemView.findViewById(R.id.tv)
            var tv2: TextView = itemView.findViewById(R.id.tv2)
            var iv: ImageView = itemView.findViewById(R.id.iv)
            var cb: CheckBox = itemView.findViewById(R.id.cb)
        }
    }
}